apis/apps/v1alpha1/clusterdefinition_types.go

```diff
@@ -945,3 +951,79 @@ func (r *ClusterDefinition) GetComponentDefByName(compDefName string) *ClusterCo
 	}
 	return nil
 }
+
+// FailurePolicyType specifies the type of failure policy
+type FailurePolicyType string
+
+const (
+	// Ignore means that an error will be ignored but logged.
+	FailurePolicyIgnore FailurePolicyType = "Ignore"
+	// ReportError means that an error will be reported.
+	FailurePolicyFail FailurePolicyType = "Fail"
+)
+
+type ComponentValueFromType string
+
+const (
+	FromFieldRef           ComponentValueFromType = "FieldRef"
+	FromServiceRef         ComponentValueFromType = "ServiceRef"
+	FromHeadlessServiceRef ComponentValueFromType = "HeadlessServiceRef"
+)
+
+// ComponentRef is used to select the component and its fields to be referenced.
+type ComponentRef struct {
+	// componentDefName is the name of the componentDef to select.
+	// +kubebuilder:validation:Required
+	ComponentDefName string `json:"componentDefName"`
+	// failurePolicy is the failure policy of the component.
+	// If failed to find the component, the failure policy will be used.
+	// +kubebuilder:validation:Enum={Ignore,Fail}
+	// +default="Ignore"
+	// +optional
+	FailurePolicy FailurePolicyType `json:"failurePolicy,omitempty"`
+	// componentRefEnv specifies a list of values to be injected as env variables to each component.
+	// +kbubebuilder:validation:Required
+	// +patchMergeKey=name
+	// +patchStrategy=merge,retainKeys
+	// +optional
+	ComponentRefEnvs []*ComponentRefEnv `json:"componentRefEnv" patchStrategy:"merge" patchMergeKey:"name"`
+}
+
+// ComponentRefEnv specifies name and value of an env.
+type ComponentRefEnv struct {
+	// name is the name of the env to be injected.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=`^[-._a-zA-Z][-._a-zA-Z0-9]*$`
+	Name string `json:"name"`
+	// value is the value of the env to be injected.
+	// +optional
+	Value string `json:"value,omitempty"`
+	// valueFrom specifies the source of the env to be injected.
+	// +optional
+	ValueFrom *ComponentValueFrom `json:"valueFrom,omitempty"`
+}
+
+type ComponentValueFrom struct {
+	// type is the type of the source to select. There are three types: `FieldRef`, `ServiceRef`, `HeadlessServiceRef`.
+	// +kubebuilder:validation:Enum={FieldRef,ServiceRef,HeadlessServiceRef}
```

Comment:
move enum marker at the type declaration

Reply:
update.